kernel/posix/mman.c

```diff
@@ -0,0 +1,598 @@
+
+	if (fchk(me->map_type, MAP_PRIVATE))
+		k_mem_slab_free(&mman_slab, &me->reg_start);
+
+	memset(me, 0, sizeof(struct _mte));
+	return 0;
+}
+
+#define map_addr_in_range(_p, _addr) \
+		((_p)->reg_start <= (_addr) && (_p)->reg_end > (_addr))
+
+inline struct _mte *map_find_by_tid_addr(k_tid_t proc, void *addr)
+{
+	struct _mte *me = NULL;
+	struct _mte *it, *lmt;
+
+	it = mt;
+	lmt = mt + MT_ENTRIES;
+
+	while (it < lmt && !me) {
+		if (it->proc == proc && map_addr_in_range(it, addr))
+			me = it;
+		++it;
+	}
+
+	return me;
+}
+
+/*
+ * Checks if the file exists, if it does, simply assign a FD to it, adding it
+ * to the FDT.
+ * If it does not, creates it and the FD.
+ */
+int shm_open(const char *name, int oflag, mode_t mode)
+{
+	struct _gfte *gf;
+	struct _fdte *fde;
+	unsigned int key;
+	bool can_write, can_read, created_file = false;
+	int used_flag, ret;
+
+	key = irq_lock();
+
+	/* This check avoids checking internally */
+	if (name == NULL)
+		SET_ERRNO_AND_RET(EINVAL);
+
+	used_flag = (fchk(oflag, O_RDONLY)) ? O_RDONLY : O_RDWR;
+
+	gf = file_find_by_name(name);
+	if (gf) {
+		if (fchk(oflag, O_EXCL))
+			SET_ERRNO_AND_RET(EEXIST);
+
+		/*
+		 * Assume there is only one user in the system
+		 */
+		can_read = fchk(gf->perm, S_IRUSR);
+		can_write = fchk(gf->perm, S_IWUSR);
+
+		if (!can_read || (!can_write && used_flag == O_RDWR))
+			SET_ERRNO_AND_RET(EACCES);
+
+		/* Truncate file */
+		if (fchk(oflag, O_TRUNC) && can_write)
+			memset(gf->buff, REGION_SIZE, 0);
+
+	} else {
+		if (!fchk(oflag, O_CREAT))
+			SET_ERRNO_AND_RET(ENOENT);
+
+		gf = file_create(name, mode);
+		if (!gf)
+			SET_ERRNO_AND_RET(ENOSPC);
+		created_file = true;
+	}
+
+	/* If this fails we need to cleanup */
+	ret = fds_create(gf, used_flag, &fde);
+	if (ret != 0) {
+		if (created_file)
+			file_delete(gf);
+		SET_ERRNO_AND_RET(ret);
+	}
+
+	irq_unlock(key);
+	return fde->fd;
+
+error:
+	irq_unlock(key);
+	return -1;
+}
+
+/*
+ * Marks memory region to release or frees it
+ */
+int shm_unlink(const char *name)
+{
+	struct _gfte *gf;
+	unsigned int key;
+
+	key = irq_lock();
+	gf = file_find_by_name(name);
+
+	if (!gf)
+		SET_ERRNO_AND_RET(ENOENT);
+
+	if (!fds_count_references(gf))
+		file_delete(gf);
+	else
+		file_unlink(gf);
+
+	irq_unlock(key);
+	return 0;
+
+error:
+	irq_unlock(key);
+	return -1;
+}
+
+/*
+ * prot is made to set page permissions in MMU, but in Zephyr we do not have
+ */
+void *mmap(void *addr, size_t len, int prot, int flags, int fildes, off_t off)
+{
+	struct _fdte *fde;
+	struct _mte *mapping;
+	unsigned int key;
+
+	key = irq_lock();
+
+	/* We don't support this */
+	if (fchk(flags, MAP_FIXED) || addr != NULL || !len ||
+	    (!fchk(flags, MAP_SHARED) && !fchk(flags, MAP_PRIVATE)) || off < 0)
+		SET_ERRNO_AND_RET(EINVAL);
+
+	fde = fds_find_by_fd(fildes);
+	if (!fde)
+		SET_ERRNO_AND_RET(EBADF);
+
+	/* TODO: Compare file descriptor permissions with protection bits */
+
+	/* Check bounds of mapping */
+	if (off + len > REGION_SIZE)
```

Comment:
It's often better to do this the other way around: `REGION_SIZE <= off - len`, as a way to work around `off + len` overflowing